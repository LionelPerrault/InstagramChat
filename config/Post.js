// HANDY METHODS FOR POST ROUTES

const db = require('./db'),
  { unlink } = require('fs'),
  { promisify } = require('util'),
  root = process.cwd()

/**
 * Returns whther user liked the post
 * @param {Number} user User ID
 * @param {Number} post Post ID
 * @returns {Boolean} Boolean
 */
const likedOrNot = async (user, post) => {
  let s = await db.query(
    'SELECT COUNT(like_id) AS l FROM likes WHERE like_by=? AND post_id=?',
    [user, post]
  )
  return db.tf(s[0].l)
}

/**
 * Returns whether user bookmarked the post
 * @param {Number} user User ID
 * @param {Number} post Post ID
 * @returns {Boolean} Boolean
 */
const bookmarkedOrNot = async (user, post) => {
  let s = await db.query(
    'SELECT COUNT(bkmrk_id) AS b FROM bookmarks WHERE bkmrk_by=? AND post_id=?',
    [user, post]
  )
  return db.tf(s[0].b)
}

/**
 * Returns whether session is the owner of post
 * @param {Number} session Session ID
 * @param {Number} post Post ID
 * @returns {Boolean} Boolean
 */
const isPostMine = async (session, post) => {
  let s = await db.query('SELECT user FROM posts WHERE post_id=?', [post])
  return s[0].user == session ? true : false
}

/**
 * Returns whether session shares post to user
 * @param {Number} post Post ID
 * @param {Number} session Session ID [share_by]
 * @param {User} user User ID [share_to]
 * @returns {Boolean} Boolean
 */
const didIShare = async (post, session, user) => {
  let s = await db.query(
    'SELECT COUNT(share_id) AS post_share FROM shares WHERE share_by=? AND share_to=? AND post_id=?',
    [session, user, post]
  )
  return db.tf(s[0].post_share)
}

/**
 * Returns tags count, likes count, ...
 * @param {Number} post_id Post ID
 * @returns {Object} Tags Count, Likes Count, ...
 */
const getCounts = async post_id => {
  let [{ tags_count }] = await db.query(
      'SELECT COUNT(post_tag_id) AS tags_count FROM post_tags WHERE post_id=?',
      [post_id]
    ),
    [{ likes_count }] = await db.query(
      'SELECT COUNT(like_id) AS likes_count FROM likes WHERE post_id=?',
      [post_id]
    ),
    [{ shares_count }] = await db.query(
      'SELECT COUNT(share_id) AS shares_count FROM shares WHERE post_id=?',
      [post_id]
    ),
    [{ comments_count }] = await db.query(
      'SELECT COUNT(comment_id) AS comments_count FROM comments WHERE post_id=?',
      [post_id]
    )

  return {
    tags_count,
    likes_count,
    shares_count,
    comments_count,
  }
}

/** Deletes a post */
const deletePost = async ({ post, when }) => {
  await db.query('DELETE FROM likes WHERE post_id=?', [post])
  await db.query('DELETE FROM post_tags WHERE post_id=?', [post])
  await db.query('DELETE FROM shares WHERE post_id=?', [post])
  await db.query('DELETE FROM bookmarks WHERE post_id=?', [post])
  await db.query('DELETE FROM notifications WHERE post_id=?', [post])
  await db.query('DELETE FROM hashtags WHERE post_id=?', [post])

  let [{ imgSrc }] = await db.query(
      'SELECT imgSrc FROM posts WHERE post_id=?',
      [post]
    ),
    comments = await db.query(
      'SELECT commentSrc, type FROM comments WHERE post_id=?',
      [post]
    ),
    deleteFile = promisify(unlink)

  await deleteFile(`${root}/dist/posts/${imgSrc}`)

  comments.map(async c => {
    if (c.type != 'text') {
      await deleteFile(`${root}/dist/comments/${c.commentSrc}`)
    }
  })
  await db.query('DELETE FROM comments WHERE post_id=?', [post])

  if (when == 'user') {
    await db.query('DELETE FROM posts WHERE post_id=?', [post])
  } else {
    await db.query('DELETE FROM posts WHERE post_id=?', [post])
  }
}

module.exports = {
  likedOrNot,
  bookmarkedOrNot,
  isPostMine,
  didIShare,
  getCounts,
  deletePost,
}
