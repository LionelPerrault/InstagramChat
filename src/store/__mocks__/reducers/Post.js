import posts from './posts'

export default {
  feed: posts,
  posts,
  tagged: posts,
  shared: [
    {
      ...posts[0],
      share_id: 445,
      share_by_username: 'takkar',
      share_time: '1480114098767',
    },
  ],
  bookmarks: [posts[0]],
  photos: [
    {
      post_id: 43,
      user: 18,
      username: 'ragnar',
      firstname: 'Ragnar',
      surname: 'Lothbrok',
      post_time: '1480114098767',
      imgsrc: 'instagram_1516528062094.jpg',
      filter: 'filter-normal',
    },
  ],
  viewPost: {
    ...posts[0],
    comments: [
      {
        comment_id: 73,
        type: 'text',
        text: 'https://regexr.com/?37i6s fffffm',
        commentSrc: '',
        comment_by: 7,
        comment_by_username: 'ghalib',
        post_id: 89,
        comment_time: '1480114098767',
      },
      {
        comment_id: 69,
        type: 'image',
        text: '',
        commentSrc: 'instagram_comment_1518972851259.jpg',
        comment_by: 8,
        comment_by_username: 'coldplay',
        post_id: 89,
        comment_time: '1480114098767',
      },
      {
        comment_id: 104,
        type: 'sticker',
        text: '',
        commentSrc: 'instagram_comment_1527447929485.jpg',
        comment_by: 24,
        comment_by_username: 'takkar',
        post_id: 43,
        comment_time: '1480114098767',
      },
    ],
  },
  likes: [
    {
      like_id: 42,
      like_by: 18,
      username: 'ragnar',
      firstname: 'Ragnar',
      surname: 'Lothbrok',
      post_id: 43,
      like_time: '1480114098767',
      isFollowing: false,
    },
  ],
  tags: [
    {
      post_tag_id: 1,
      post_id: 89,
      user: 16,
      username: 'zayn',
      firstname: 'Zayn',
      surname: 'Malik',
      isFollowing: false,
    },
  ],
  isPostMine: false,
  usersToShare: [
    {
      follow_id: 302,
      follow_to: 16,
      username: 'zayn',
      firstname: 'Zayn',
      surname: 'Malik',
      didIShare: false,
    },
    {
      follow_id: 289,
      follow_to: 10,
      username: 'noddy',
      firstname: 'your',
      surname: 'noddy',
      didIShare: false,
    },
  ],
  sharers: [
    {
      share_id: 103,
      share_by: 7,
      share_by_username: 'ghalib',
      share_by_firstname: 'Mirza',
      share_by_surname: 'Ghalib',
      share_to: 28,
      share_to_username: 'selena',
      share_time: '1480114098767',
      post_id: 43,
      isFollowing: false,
    },
  ],
  postIt: {
    fileInput: '', // file input value
    fileChanged: false, // for checking file has changed
    targetFile: '', // file
    previewImg: '/images/location.jpg', // image which will be previewd
    desc: '', // textarea value
    filter: 'filter-normal',
    fetchingLocation: false,
    location: 'A-301, 90 Feet Road, Dharavi, Mumbai, Maharashtra 400017, India',
    addTag: false,
    tags: [
      {
        user: 7,
        username: 'ghalib',
      },
    ],
    showOverlay: false,
    type: 'user',
    group: 0,
  },
}
