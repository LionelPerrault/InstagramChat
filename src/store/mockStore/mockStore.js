import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import promise from 'redux-promise-middleware'

// mock-reducers
import User from './mock-reducers/User'
import Notification from './mock-reducers/Notification'
import Group from './mock-reducers/Group'
import Follow from './mock-reducers/Follow'
import Explore from './mock-reducers/Explore'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore
}

const middlewares = [
  promise(), thunk, logger
]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
