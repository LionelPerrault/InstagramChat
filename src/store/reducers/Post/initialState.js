import PostItInitialState from './postit-initial-state'

export default {
  posts: [],
  bookmarks: [],
  tagged: [],
  shared: [],
  photos: [],
  feed: [],
  viewPost: {},
  likes: [],
  tags: [],
  posted: false,
  isPostMine: false,
  usersToShare: [],
  sharers: [],
  postIt: PostItInitialState,
}
