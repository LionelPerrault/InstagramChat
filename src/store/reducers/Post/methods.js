import PostItIntialState from './postit-initial-state'

export const addPost = (posts, post) => {
  posts = [post, ...posts]
  return posts
}

export const editPost = (posts, { post_id, description }) => {
  return posts.map(p => {
    if (p.post_id == post_id) {
      p.description = description
    }
    return p
  })
}

export const deletePost = (posts, post) =>
  posts.filter(p => p.post_id != parseInt(post))

export const untag = (tags, user) => {
  let tt = tags.filter(t => t.user != user)
  return tt
}

export const unbookmark = (bookmarks, post) =>
  bookmarks.filter(b => b.post_id != post)

export const removeShare = (sharers, share_id) =>
  sharers.filter(s => s.share_id != share_id)

export const comment = (post, comment) => {
  post = {
    ...post,
    comments: [comment, ...post.comments],
  }
  return post
}

export const deleteComment = (post, comment_id) => {
  let comments = post.comments.filter(c => c.comment_id != comment_id)
  return {
    ...post,
    comments,
  }
}

export const editComment = (post, { comment_id, comment }) => {
  let comments = post.comments.map(c => {
    if (c.comment_id == comment_id) {
      c.text = comment
    }
    return c
  })

  return {
    ...post,
    comments,
  }
}

export const removeLike = (likes, like_id) =>
  likes.filter(l => l.like_id != like_id)

export const changePostIt = (postIt, pyOptions) => {
  let { what, value } = pyOptions
  let updated = {
    ...postIt,
    [what]: value,
  }
  return updated
}

export const resetPostItProperties = () => {
  let reset = PostItIntialState
  return reset
}
