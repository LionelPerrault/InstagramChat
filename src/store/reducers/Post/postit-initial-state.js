export default {
  fileInput: '', // file input value
  fileChanged: false, // for checking file has changed
  targetFile: '', // file
  previewImg: '/images/location.jpg', // image which will be previewd
  desc: '', // textarea value
  filter: 'filter-normal',
  fetchingLocation: false,
  location: '',
  addTag: false,
  tags: [],
  showOverlay: false,
  type: '',
  group: null,
}
