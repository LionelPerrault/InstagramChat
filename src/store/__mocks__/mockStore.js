import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

// mock-reducers
import User from './reducers/User'
import Notification from './reducers/Notification'
import Group from './reducers/Group'
import Follow from './reducers/Follow'
import Explore from './reducers/Explore'
import Post from './reducers/Post'
import Hashtag from './reducers/Hashtag'
import Setting from './reducers/Setting'
import Message from './reducers/Message'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore,
  Post,
  Hashtag,
  Setting,
  Message,
}

const middlewares = [thunk]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
