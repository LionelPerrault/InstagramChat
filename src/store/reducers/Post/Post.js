/* eslint indent:0 */
/* eslint no-unreachable:0 */

import initialState from './initialState'
import * as methods from './methods'

export default (state = initialState, action) => {
  let py = action.payload

  switch (action.type) {
    case 'GET_USER_POSTS':
      return { ...state, posts: py }
      break

    case 'GET_BOOKMARKED_POSTS':
      return { ...state, bookmarks: py }
      break

    case 'GET_TAGGED_POSTS':
      return { ...state, tagged: py }
      break

    case 'GET_SHARED_POSTS':
      return { ...state, shared: py }
      break

    case 'GET_PHOTOS':
      return { ...state, photos: py }
      break

    case 'GET_FEED':
      return { ...state, feed: py }
      break

    case 'GET_GROUP_POSTS':
      return { ...state, posts: py }
      break

    case 'GET_GROUP_PHOTOS':
      return { ...state, photos: py }
      break

    case 'GET_POST':
      return { ...state, viewPost: py }
      break

    case 'ADD_USER_POST':
      return {
        ...state,
        feed: methods.addPost(state.feed, py),
      }
      break

    case 'ADD_GROUP_POST':
      return {
        ...state,
        posts: methods.addPost(state.posts, py),
      }
      break

    case 'EDIT_POST':
      return {
        ...state,
        posts: methods.editPost(state.posts, py),
      }
      break

    case 'DELETE_POST':
      return {
        ...state,
        posts: methods.deletePost(state.posts, py),
        feed: methods.deletePost(state.feed, py),
      }
      break

    case 'GET_POST_LIKES':
      return {
        ...state,
        likes: py.likes,
        isPostMine: py.isPostMine,
      }
      break

    case 'REMOVE_LIKE':
      return {
        ...state,
        likes: methods.removeLike(state.likes, py),
      }
      break

    case 'GET_POST_TAGS':
      return {
        ...state,
        tags: py.tags,
        isPostMine: py.isPostMine,
      }
      break

    case 'UNTAG':
      return {
        ...state,
        tags: methods.untag(state.tags, py),
      }
      break

    case 'GET_USERS_TO_SHARE':
      return { ...state, usersToShare: py }
      break

    case 'GET_POST_SHARERS':
      return { ...state, sharers: py }
      break

    case 'UNBOOKMARK':
      return {
        ...state,
        bookmarks: methods.unbookmark(state.bookmarks, py),
      }
      break

    case 'REMOVE_SHARE':
      return {
        ...state,
        sharers: methods.removeShare(state.sharers, py),
      }
      break

    case 'COMMENT':
      return {
        ...state,
        viewPost: methods.comment(state.viewPost, py),
      }
      break

    case 'DELETE_COMMENT':
      return {
        ...state,
        viewPost: methods.deleteComment(state.viewPost, py),
      }
      break

    case 'EDIT_COMMENT':
      return {
        ...state,
        viewPost: methods.editComment(state.viewPost, py),
      }
      break

    // Post It
    case 'CHANGE_POSTIT_PROPS':
      return {
        ...state,
        postIt: methods.changePostIt(state.postIt, py),
      }

    case 'RESET_POSTIT':
      return {
        ...state,
        postIt: methods.resetPostItProperties(),
      }
  }

  return state
}
