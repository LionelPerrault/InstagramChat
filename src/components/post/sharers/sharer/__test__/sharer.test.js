import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import mockStore from '../../../../../store/__mocks__/mockStore'
import Sharer from '../sharer'
import Post from '../../../../../store/__mocks__/reducers/Post'
import MockDataElement from '../../../../../utils/__mocks__/mock-dataElement'

describe('Sharer Component', () => {
  let dataElement

  beforeEach(() => (dataElement = MockDataElement()))

  afterEach(() => dataElement.remove())

  const comp = (extraProps = {}) => (
    <Provider store={mockStore}>
      <Sharer
        {...Post.sharers[0]}
        decrementSharers={jest.fn()}
        {...extraProps}
      />
    </Provider>
  )

  it('should match snapshot and show <AdvancedFollow/>', () => {
    const tree = create(comp()).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot and show <AdvancedUnfollow/>', () => {
    const tree = create(
      comp({
        isFollowing: true,
      })
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot and show <AppLink/>', () => {
    const tree = create(
      comp({
        share_by: 24,
      })
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot and show <RemoveShare/>', () => {
    dataElement.setAttribute('data-isadmin', 'true')
    const tree = create(comp()).toJSON()
    expect(tree).toMatchSnapshot()

    dataElement.setAttribute('data-isadmin', 'false')
    const tree2 = create(
      comp({
        share_to: 24,
      })
    ).toJSON()
    expect(tree2).toMatchSnapshot()
  })
})
