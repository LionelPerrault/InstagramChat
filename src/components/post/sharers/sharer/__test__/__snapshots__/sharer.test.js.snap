// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sharer Component should match snapshot and show <AdvancedFollow/> 1`] = `
<div
  className="modal_items fer_items"
>
  <div
    className="modal_it_img"
  >
    <img
      src="/users/7/avatar.jpg"
    />
  </div>
  <div
    className="modal_it_content"
  >
    <div
      className="modal_it_info"
    >
      <Link
        className="modal_it_username"
        to="/profile/ghalib"
      >
        ghalib
      </Link>
      <span
        className="modal_it_light"
      >
        to 
        selena
      </span>
    </div>
    <div
      className="modal_ff"
    >
      <a
        className="pri_btn follow"
        href="#"
        onClick={[Function]}
      >
        Follow
      </a>
    </div>
  </div>
  <hr />
</div>
`;

exports[`Sharer Component should match snapshot and show <AdvancedUnfollow/> 1`] = `
<div
  className="modal_items fer_items"
>
  <div
    className="modal_it_img"
  >
    <img
      src="/users/7/avatar.jpg"
    />
  </div>
  <div
    className="modal_it_content"
  >
    <div
      className="modal_it_info"
    >
      <Link
        className="modal_it_username"
        to="/profile/ghalib"
      >
        ghalib
      </Link>
      <span
        className="modal_it_light"
      >
        to 
        selena
      </span>
    </div>
    <div
      className="modal_ff"
    >
      <a
        className="pri_btn unfollow"
        href="#"
        onClick={[Function]}
      >
        Unfollow
      </a>
    </div>
  </div>
  <hr />
</div>
`;

exports[`Sharer Component should match snapshot and show <AppLink/> 1`] = `
<div
  className="modal_items fer_items"
>
  <div
    className="modal_it_img"
  >
    <img
      src="/users/24/avatar.jpg"
    />
  </div>
  <div
    className="modal_it_content"
  >
    <div
      className="modal_it_info"
    >
      <Link
        className="modal_it_username"
        to="/profile/ghalib"
      >
        You
      </Link>
      <span
        className="modal_it_light"
      >
        to 
        selena
      </span>
    </div>
    <div
      className="modal_ff"
    >
      <Link
        className="pri_btn follow"
        to="/profile/ghalib"
      >
        Profile
      </Link>
    </div>
  </div>
  <hr />
</div>
`;

exports[`Sharer Component should match snapshot and show <RemoveShare/> 1`] = `
<div
  className="modal_items fer_items"
>
  <div
    className="modal_it_img"
  >
    <img
      src="/users/7/avatar.jpg"
    />
  </div>
  <div
    className="modal_it_content"
  >
    <div
      className="modal_it_info"
    >
      <Link
        className="modal_it_username"
        to="/profile/ghalib"
      >
        ghalib
      </Link>
      <span
        className="modal_it_light"
      >
        to 
        selena
      </span>
    </div>
    <div
      className="modal_ff"
    >
      <a
        className="sec_btn"
        href="#"
        onClick={[Function]}
      >
        Remove as admin
      </a>
    </div>
  </div>
  <hr />
</div>
`;

exports[`Sharer Component should match snapshot and show <RemoveShare/> 2`] = `
<div
  className="modal_items fer_items"
>
  <div
    className="modal_it_img"
  >
    <img
      src="/users/7/avatar.jpg"
    />
  </div>
  <div
    className="modal_it_content"
  >
    <div
      className="modal_it_info"
    >
      <Link
        className="modal_it_username"
        to="/profile/ghalib"
      >
        ghalib
      </Link>
      <span
        className="modal_it_light"
      >
        to 
        You
      </span>
    </div>
    <div
      className="modal_ff"
    >
      <a
        className="sec_btn"
        href="#"
        onClick={[Function]}
      >
        Remove share
      </a>
    </div>
  </div>
  <hr />
</div>
`;
